corehq/util/metrics/tests/test_metrics.py

```diff
@@ -0,0 +1,171 @@
+from django.test import SimpleTestCase
+
+from corehq.util.metrics import DatadogMetrics, PrometheusMetrics
+from corehq.util.metrics.metrics import HqCounter, HqGauge, HqHistogram
+from corehq.util.metrics.tests.utils import patch_datadog
+from prometheus_client.samples import Sample
+from prometheus_client.utils import INF
+
+
+class _TestMetrics(SimpleTestCase):
+    provider_class = None
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.provider = cls.provider_class()
+
+    def test_counter(self):
+        counter = self.provider.counter('commcare.test.counter', 'Description', tag_names=['t1', 't2'])
+        counter.tag(t1='a', t2='b').inc()
+        counter.tag(t1='c', t2='b').inc(2)
+        counter.tag(t1='c', t2='b').inc()
+        self.assertCounterMetric(counter, {
+            (('t1', 'a'), ('t2', 'b')): 1,
+            (('t1', 'c'), ('t2', 'b')): 3,
+        })
+
+    def test_gauge(self):
+        gauge = self.provider.gauge('commcare.test.gauge', 'Description', tag_names=['t1', 't2'])
+        gauge.tag(t1='a', t2='b').set(4.2)
+        gauge.tag(t1='c', t2='b').set(2)
+        gauge.tag(t1='c', t2='b').set(5)
+        self.assertGaugeMetric(gauge, {
+            (('t1', 'a'), ('t2', 'b')): 4.2,
+            (('t1', 'c'), ('t2', 'b')): 5,
+        })
+
+    def test_histogram(self):
+        histogram = self.provider.histogram(
+            'commcare.test.histogram', 'Description', 'duration',
+            buckets=[1, 2, 3], bucket_unit='ms', tag_names=['t1', 't2']
+        )
+        tagged_1 = histogram.tag(t1='a', t2='b')
+        tagged_1.observe(0.2)
+        tagged_1.observe(0.7)
+        tagged_1.observe(2.5)
+
+        tagged_2 = histogram.tag(t1='c', t2='b')
+        tagged_2.observe(2)
+        tagged_2.observe(5)
+        self.assertHistogramMetric(histogram, {
+            (('t1', 'a'), ('t2', 'b')): {1: 2, 3: 1},
+            (('t1', 'c'), ('t2', 'b')): {2: 1, INF: 1}
+        })
+
+    def assertCounterMetric(self, metric: HqCounter, expected: Dict[Tuple[Tuple[str, str], ...], float]):
+        """
+        :param metric: metric class
+        :param expected: dict mapping tag tuples to metric values
+        """
+        raise NotImplementedError
+
+    def assertGaugeMetric(self, metric: HqGauge, expected: Dict[Tuple[Tuple[str, str], ...], float]):
+        """
+        :param metric: metric class
+        :param expected: dict mapping tag tuples to metric values
+        """
+        raise NotImplementedError
+
+    def assertHistogramMetric(self,
+                              metric: HqHistogram,
+                              expected: Dict[Tuple[Tuple[str, str], ...], Dict[float, int]]):
+        """
+        :param metric: metric class
+        :param expected: dict mapping tag tuples to a dict of bucket values
+        """
+        raise NotImplementedError
+
+
+class TestDatadogMetrics(_TestMetrics):
+    provider_class = DatadogMetrics
+
+    def setUp(self) -> None:
```

Comment:
Do type annotations like this do anything useful?

Reply:
No, it should be removed. This was added by autocomplete in pycharm and I never removed it.